{
 "cells": [
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-17T08:43:55.312706Z",
     "start_time": "2025-04-17T08:43:55.310854Z"
    }
   },
   "source": [],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {
    "scrolled": true,
    "ExecuteTime": {
     "end_time": "2025-04-17T08:43:55.362248Z",
     "start_time": "2025-04-17T08:43:55.335244Z"
    }
   },
   "source": [
    "from dash import Dash, html, dcc, dash_table, Input, Output, callback_context\n",
    "import plotly.express as px\n",
    "import pandas as pd\n",
    "from animal_shelter import AnimalShelter\n",
    "\n",
    "# Create instance of the AnimalShelter class\n",
    "shelter = AnimalShelter(\"myUserAdmin\", \"trinidad1\")\n",
    "\n",
    "# Initialize the Dash app\n",
    "app = Dash(__name__)\n",
    "\n",
    "# Create the layout\n",
    "app.layout = html.Div([\n",
    "    # Title\n",
    "    html.Center(html.B(html.H1('SNHU CS-340 Dashboard'))),\n",
    "\n",
    "    # Filter buttons\n",
    "    html.Center(html.Div([\n",
    "        html.Button('Cats', id='cat-button', n_clicks=0),\n",
    "        html.Button('Dogs', id='dog-button', n_clicks=0),\n",
    "        html.Button('Reset', id='reset-button', n_clicks=0)\n",
    "    ])),\n",
    "\n",
    "    # Data table\n",
    "    dash_table.DataTable(\n",
    "        id='datatable',\n",
    "        columns=[\n",
    "            {\"name\": i, \"id\": i} for i in ['animal_type', 'breed', 'color', 'age_upon_outcome']\n",
    "        ],\n",
    "        page_size=10\n",
    "    ),\n",
    "\n",
    "    # Graph\n",
    "    dcc.Graph(id='data-graph')\n",
    "])\n",
    "\n",
    "# Callback to update dashboard\n",
    "@app.callback(\n",
    "    [Output('datatable', 'data'),\n",
    "     Output('data-graph', 'figure')],\n",
    "    [Input('cat-button', 'n_clicks'),\n",
    "     Input('dog-button', 'n_clicks'),\n",
    "     Input('reset-button', 'n_clicks')]\n",
    ")\n",
    "def update_dashboard(cat_clicks, dog_clicks, reset_clicks):\n",
    "    ctx = callback_context\n",
    "    if not ctx.triggered:\n",
    "        query = {}\n",
    "    else:\n",
    "        button_id = ctx.triggered[0]['prop_id'].split('.')[0]\n",
    "        if button_id == \"cat-button\":\n",
    "            query = {\"animal_type\": \"Cat\"}\n",
    "        elif button_id == \"dog-button\":\n",
    "            query = {\"animal_type\": \"Dog\"}\n",
    "        else:\n",
    "            query = {}\n",
    "\n",
    "    # Get data\n",
    "    cursor = shelter.read_all(query)\n",
    "    df = pd.DataFrame(list(cursor))\n",
    "\n",
    "    # Create visualization\n",
    "    if not df.empty:\n",
    "        fig = px.bar(df['breed'].value_counts().head(10),\n",
    "                     title='Top 10 Breeds')\n",
    "    else:\n",
    "        fig = px.bar(title='No Data Available')\n",
    "\n",
    "    return df.to_dict('records'), fig\n",
    "\n",
    "# Run the app\n",
    "if __name__ == '__main__':\n",
    "    app.run(debug=True)"
   ],
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<IPython.lib.display.IFrame at 0x1245696d0>"
      ],
      "text/html": [
       "\n",
       "        <iframe\n",
       "            width=\"100%\"\n",
       "            height=\"650\"\n",
       "            src=\"http://127.0.0.1:8050/\"\n",
       "            frameborder=\"0\"\n",
       "            allowfullscreen\n",
       "            \n",
       "        ></iframe>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "---------------------------------------------------------------------------\n",
      "ServerSelectionTimeoutError               Traceback (most recent call last)\n",
      "Cell In[7], line 60, in update_dashboard(\n",
      "    cat_clicks=0,\n",
      "    dog_clicks=0,\n",
      "    reset_clicks=0\n",
      ")\n",
      "     58 # Get data\n",
      "     59 cursor = shelter.read_all(query)\n",
      "---> 60 df = pd.DataFrame(list(cursor))\n",
      "        cursor = <pymongo.synchronous.cursor.Cursor object at 0x12520a100>\n",
      "        df =       Unnamed: 0 age_upon_outcome animal_id animal_type  \\\n",
      "0              1          3 years   A746874         Cat   \n",
      "1              2           1 year   A725717         Cat   \n",
      "2              3          2 years   A716330         Dog   \n",
      "3              4         7 months   A733653         Cat   \n",
      "4              5          2 years   A691584         Dog   \n",
      "...          ...              ...       ...         ...   \n",
      "9995        9996          3 weeks   A678494         Cat   \n",
      "9996        9997          5 years   A745429         Cat   \n",
      "9997        9998         2 months   A697165         Dog   \n",
      "9998        9999         2 months   A665304         Dog   \n",
      "9999       10000          4 weeks   A728994       Other   \n",
      "\n",
      "                                   breed         color date_of_birth  \\\n",
      "0                 Domestic Shorthair Mix   Black/White    2014-04-10   \n",
      "1                 Domestic Shorthair Mix  Silver Tabby    2015-05-02   \n",
      "2                Chihuahua Shorthair Mix   Brown/White    2013-11-18   \n",
      "3                            Siamese Mix    Seal Point    2016-01-25   \n",
      "4                 Labrador Retriever Mix     Tan/White    2012-11-06   \n",
      "...                                  ...           ...           ...   \n",
      "9995                         Siamese Mix  Tortie Point    2014-04-17   \n",
      "9996              Domestic Shorthair Mix         Black    2012-03-17   \n",
      "9997  Labrador Retriever/Anatol Shepherd   White/Black    2015-02-18   \n",
      "9998  Newfoundland/Australian Cattle Dog   Black/White    2013-08-16   \n",
      "9999                                 Bat           Red    2016-05-12   \n",
      "\n",
      "                 datetime            monthyear   name outcome_subtype  \\\n",
      "0     2017-04-11 09:00:00  2017-04-11T09:00:00    NaN            SCRP   \n",
      "1     2016-05-06 10:49:00  2016-05-06T10:49:00    NaN            SCRP   \n",
      "2     2015-12-28 18:43:00  2015-12-28T18:43:00  Frank             NaN   \n",
      "3     2016-08-27 18:11:00  2016-08-27T18:11:00  Kitty             NaN   \n",
      "4     2015-05-30 13:48:00  2015-05-30T13:48:00   Luke             NaN   \n",
      "...                   ...                  ...    ...             ...   \n",
      "9995  2014-05-09 15:28:00  2014-05-09T15:28:00    NaN         Partner   \n",
      "9996  2017-03-22 12:08:00  2017-03-22T12:08:00    NaN         Partner   \n",
      "9997  2015-05-03 00:00:00  2015-05-03T00:00:00  *Zoey         Offsite   \n",
      "9998  2013-10-20 17:46:00  2013-10-20T17:46:00  Flynn             NaN   \n",
      "9999  2016-06-12 12:48:00  2016-06-12T12:48:00    NaN     Rabies Risk   \n",
      "\n",
      "         outcome_type sex_upon_outcome  location_lat  location_long  \\\n",
      "0            Transfer    Neutered Male     30.506658     -97.340878   \n",
      "1            Transfer    Spayed Female     30.652598     -97.741996   \n",
      "2            Adoption    Neutered Male     30.759575     -97.552375   \n",
      "3            Adoption    Intact Female     30.318806     -97.724038   \n",
      "4     Return to Owner    Neutered Male     30.710482     -97.562297   \n",
      "...               ...              ...           ...            ...   \n",
      "9995         Transfer    Intact Female     30.718965     -97.705055   \n",
      "9996         Transfer    Intact Female     30.302056     -97.325988   \n",
      "9997         Adoption    Spayed Female     30.477896     -97.356929   \n",
      "9998         Adoption    Neutered Male     30.382226     -97.321557   \n",
      "9999       Euthanasia          Unknown     30.666477     -97.403375   \n",
      "\n",
      "      age_upon_outcome_in_weeks  \n",
      "0                    156.767857  \n",
      "1                     52.921528  \n",
      "2                    110.111409  \n",
      "3                     30.822520  \n",
      "4                    133.653571  \n",
      "...                         ...  \n",
      "9995                   3.234921  \n",
      "9996                 261.643651  \n",
      "9997                  10.571429  \n",
      "9998                   9.391468  \n",
      "9999                   4.504762  \n",
      "\n",
      "[10000 rows x 16 columns]\n",
      "        pd = <module 'pandas' from '/Users/jadathibodeaux/Library/Python/3.9/lib/python/site-packages/pandas/__init__.py'>\n",
      "     62 # Create visualization\n",
      "     63 if not df.empty:\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/cursor.py:1284, in Cursor.__next__(\n",
      "    self=<pymongo.synchronous.cursor.Cursor object>\n",
      ")\n",
      "   1283 def __next__(self) -> _DocumentType:\n",
      "-> 1284     return self.next()\n",
      "        self = <pymongo.synchronous.cursor.Cursor object at 0x12520a100>\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/cursor.py:1260, in Cursor.next(\n",
      "    self=<pymongo.synchronous.cursor.Cursor object>\n",
      ")\n",
      "   1258 if self._empty:\n",
      "   1259     raise StopIteration\n",
      "-> 1260 if len(self._data) or self._refresh():\n",
      "        self = <pymongo.synchronous.cursor.Cursor object at 0x12520a100>\n",
      "        len(self._data) = 0\n",
      "        self._data = deque([])\n",
      "   1261     return self._data.popleft()\n",
      "   1262 else:\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/cursor.py:1208, in Cursor._refresh(\n",
      "    self=<pymongo.synchronous.cursor.Cursor object>\n",
      ")\n",
      "   1186         raise InvalidOperation(\n",
      "   1187             \"Passing a 'hint' is required when using the min/max query\"\n",
      "   1188             \" option to ensure the query utilizes the correct index\"\n",
      "   1189         )\n",
      "   1190     q = self._query_class(\n",
      "   1191         self._query_flags,\n",
      "   1192         self._collection.database.name,\n",
      "   (...)\n",
      "   1206         self._exhaust,\n",
      "   1207     )\n",
      "-> 1208     self._send_message(q)\n",
      "        q = <pymongo.message._Query object at 0x1247489e0>\n",
      "        self = <pymongo.synchronous.cursor.Cursor object at 0x12520a100>\n",
      "   1209 elif self._id:  # Get More\n",
      "   1210     if self._limit:\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/cursor.py:1102, in Cursor._send_message(\n",
      "    self=<pymongo.synchronous.cursor.Cursor object>,\n",
      "    operation=<pymongo.message._Query object>\n",
      ")\n",
      "   1099     raise InvalidOperation(\"exhaust cursors do not support auto encryption\")\n",
      "   1101 try:\n",
      "-> 1102     response = client._run_operation(\n",
      "        client = MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True)\n",
      "        self._address = None\n",
      "        self = <pymongo.synchronous.cursor.Cursor object at 0x12520a100>\n",
      "        operation = <pymongo.message._Query object at 0x1247489e0>\n",
      "   1103         operation, self._unpack_response, address=self._address\n",
      "   1104     )\n",
      "   1105 except OperationFailure as exc:\n",
      "   1106     if exc.code in _CURSOR_CLOSED_ERRORS or self._exhaust:\n",
      "   1107         # Don't send killCursors because the cursor is already closed.\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/_csot.py:119, in apply.<locals>.csot_wrapper(\n",
      "    self=MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True),\n",
      "    *args=(<pymongo.message._Query object>, <bound method Cursor._unpack_response of <pymongo.synchronous.cursor.Cursor object>>),\n",
      "    **kwargs={'address': None}\n",
      ")\n",
      "    117         with _TimeoutContext(timeout):\n",
      "    118             return func(self, *args, **kwargs)\n",
      "--> 119 return func(self, *args, **kwargs)\n",
      "        func = <function MongoClient._run_operation at 0x121698280>\n",
      "        self = MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True)\n",
      "        args = (<pymongo.message._Query object at 0x1247489e0>, <bound method Cursor._unpack_response of <pymongo.synchronous.cursor.Cursor object at 0x12520a100>>)\n",
      "        kwargs = {'address': None}\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/mongo_client.py:1896, in MongoClient._run_operation(\n",
      "    self=MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True),\n",
      "    operation=<pymongo.message._Query object>,\n",
      "    unpack_res=<bound method Cursor._unpack_response of <pymongo.synchronous.cursor.Cursor object>>,\n",
      "    address=None\n",
      ")\n",
      "   1886     operation.reset()  # Reset op in case of retry.\n",
      "   1887     return server.run_operation(\n",
      "   1888         conn,\n",
      "   1889         operation,\n",
      "   (...)\n",
      "   1893         self,\n",
      "   1894     )\n",
      "-> 1896 return self._retryable_read(\n",
      "        self = MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True)\n",
      "        operation = <pymongo.message._Query object at 0x1247489e0>\n",
      "        operation.read_preference = Primary()\n",
      "        operation.session = <pymongo.synchronous.client_session.ClientSession object at 0x12448cdc0>\n",
      "        address = None\n",
      "        operation.name = 'find'\n",
      "   1897     _cmd,\n",
      "   1898     operation.read_preference,\n",
      "   1899     operation.session,  # type: ignore[arg-type]\n",
      "   1900     address=address,\n",
      "   1901     retryable=isinstance(operation, _Query),\n",
      "   1902     operation=operation.name,\n",
      "   1903 )\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/mongo_client.py:2005, in MongoClient._retryable_read(\n",
      "    self=MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True),\n",
      "    func=<function MongoClient._run_operation.<locals>._cmd>,\n",
      "    read_pref=Primary(),\n",
      "    session=<pymongo.synchronous.client_session.ClientSession object>,\n",
      "    operation='find',\n",
      "    address=None,\n",
      "    retryable=True,\n",
      "    operation_id=None\n",
      ")\n",
      "   2000 # Ensure that the client supports retrying on reads and there is no session in\n",
      "   2001 # transaction, otherwise, we will not support retry behavior for this call.\n",
      "   2002 retryable = bool(\n",
      "   2003     retryable and self.options.retry_reads and not (session and session.in_transaction)\n",
      "   2004 )\n",
      "-> 2005 return self._retry_internal(\n",
      "        self = MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True)\n",
      "        func = <function MongoClient._run_operation.<locals>._cmd at 0x12474c310>\n",
      "        session = <pymongo.synchronous.client_session.ClientSession object at 0x12448cdc0>\n",
      "        operation = 'find'\n",
      "        address = None\n",
      "        read_pref = Primary()\n",
      "        retryable = True\n",
      "        operation_id = None\n",
      "   2006     func,\n",
      "   2007     session,\n",
      "   2008     None,\n",
      "   2009     operation,\n",
      "   2010     is_read=True,\n",
      "   2011     address=address,\n",
      "   2012     read_pref=read_pref,\n",
      "   2013     retryable=retryable,\n",
      "   2014     operation_id=operation_id,\n",
      "   2015 )\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/_csot.py:119, in apply.<locals>.csot_wrapper(\n",
      "    self=MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True),\n",
      "    *args=(<function MongoClient._run_operation.<locals>._cmd>, <pymongo.synchronous.client_session.ClientSession object>, None, 'find'),\n",
      "    **kwargs={'address': None, 'is_read': True, 'operation_id': None, 'read_pref': Primary(), 'retryable': True}\n",
      ")\n",
      "    117         with _TimeoutContext(timeout):\n",
      "    118             return func(self, *args, **kwargs)\n",
      "--> 119 return func(self, *args, **kwargs)\n",
      "        func = <function MongoClient._retry_internal at 0x121698430>\n",
      "        self = MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True)\n",
      "        args = (<function MongoClient._run_operation.<locals>._cmd at 0x12474c310>, <pymongo.synchronous.client_session.ClientSession object at 0x12448cdc0>, None, 'find')\n",
      "        kwargs = {'is_read': True, 'address': None, 'read_pref': Primary(), 'retryable': True, 'operation_id': None}\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/mongo_client.py:1961, in MongoClient._retry_internal(\n",
      "    self=MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True),\n",
      "    func=<function MongoClient._run_operation.<locals>._cmd>,\n",
      "    session=<pymongo.synchronous.client_session.ClientSession object>,\n",
      "    bulk=None,\n",
      "    operation='find',\n",
      "    is_read=True,\n",
      "    address=None,\n",
      "    read_pref=Primary(),\n",
      "    retryable=True,\n",
      "    operation_id=None\n",
      ")\n",
      "   1935 @_csot.apply\n",
      "   1936 def _retry_internal(\n",
      "   1937     self,\n",
      "   (...)\n",
      "   1946     operation_id: Optional[int] = None,\n",
      "   1947 ) -> T:\n",
      "   1948     \"\"\"Internal retryable helper for all client transactions.\n",
      "   1949 \n",
      "   1950     :param func: Callback function we want to retry\n",
      "   (...)\n",
      "   1959     :return: Output of the calling func()\n",
      "   1960     \"\"\"\n",
      "-> 1961     return _ClientConnectionRetryable(\n",
      "        self = MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True)\n",
      "        func = <function MongoClient._run_operation.<locals>._cmd at 0x12474c310>\n",
      "        bulk = None\n",
      "        operation = 'find'\n",
      "        is_read = True\n",
      "        session = <pymongo.synchronous.client_session.ClientSession object at 0x12448cdc0>\n",
      "        read_pref = Primary()\n",
      "        address = None\n",
      "        retryable = True\n",
      "        operation_id = None\n",
      "   1962         mongo_client=self,\n",
      "   1963         func=func,\n",
      "   1964         bulk=bulk,\n",
      "   1965         operation=operation,\n",
      "   1966         is_read=is_read,\n",
      "   1967         session=session,\n",
      "   1968         read_pref=read_pref,\n",
      "   1969         address=address,\n",
      "   1970         retryable=retryable,\n",
      "   1971         operation_id=operation_id,\n",
      "   1972     ).run()\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/mongo_client.py:2709, in _ClientConnectionRetryable.run(\n",
      "    self=<pymongo.synchronous.mongo_client._ClientConnectionRetryable object>\n",
      ")\n",
      "   2707 self._check_last_error(check_csot=True)\n",
      "   2708 try:\n",
      "-> 2709     return self._read() if self._is_read else self._write()\n",
      "        self = <pymongo.synchronous.mongo_client._ClientConnectionRetryable object at 0x12448c7c0>\n",
      "        self._is_read = True\n",
      "   2710 except ServerSelectionTimeoutError:\n",
      "   2711     # The application may think the write was never attempted\n",
      "   2712     # if we raise ServerSelectionTimeoutError on the retry\n",
      "   2713     # attempt. Raise the original exception instead.\n",
      "   2714     self._check_last_error()\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/mongo_client.py:2854, in _ClientConnectionRetryable._read(\n",
      "    self=<pymongo.synchronous.mongo_client._ClientConnectionRetryable object>\n",
      ")\n",
      "   2849 def _read(self) -> T:\n",
      "   2850     \"\"\"Wrapper method for read-type retryable client executions\n",
      "   2851 \n",
      "   2852     :return: Output for func()'s call\n",
      "   2853     \"\"\"\n",
      "-> 2854     self._server = self._get_server()\n",
      "        self._server = None\n",
      "        self = <pymongo.synchronous.mongo_client._ClientConnectionRetryable object at 0x12448c7c0>\n",
      "   2855     assert self._read_pref is not None, \"Read Preference required on read calls\"\n",
      "   2856     with self._client._conn_from_server(self._read_pref, self._server, self._session) as (\n",
      "   2857         conn,\n",
      "   2858         read_pref,\n",
      "   2859     ):\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/mongo_client.py:2802, in _ClientConnectionRetryable._get_server(\n",
      "    self=<pymongo.synchronous.mongo_client._ClientConnectionRetryable object>\n",
      ")\n",
      "   2797 def _get_server(self) -> Server:\n",
      "   2798     \"\"\"Retrieves a server object based on provided object context\n",
      "   2799 \n",
      "   2800     :return: Abstraction to connect to server\n",
      "   2801     \"\"\"\n",
      "-> 2802     return self._client._select_server(\n",
      "        self._client = MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True)\n",
      "        self = <pymongo.synchronous.mongo_client._ClientConnectionRetryable object at 0x12448c7c0>\n",
      "        self._server_selector = Primary()\n",
      "        self._session = <pymongo.synchronous.client_session.ClientSession object at 0x12448cdc0>\n",
      "        self._operation = 'find'\n",
      "        self._address = None\n",
      "        self._deprioritized_servers = []\n",
      "        self._operation_id = None\n",
      "   2803         self._server_selector,\n",
      "   2804         self._session,\n",
      "   2805         self._operation,\n",
      "   2806         address=self._address,\n",
      "   2807         deprioritized_servers=self._deprioritized_servers,\n",
      "   2808         operation_id=self._operation_id,\n",
      "   2809     )\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/mongo_client.py:1791, in MongoClient._select_server(\n",
      "    self=MongoClient(host=['localhost:54319'], document_class=dict, tz_aware=False, connect=True),\n",
      "    server_selector=Primary(),\n",
      "    session=<pymongo.synchronous.client_session.ClientSession object>,\n",
      "    operation='find',\n",
      "    address=None,\n",
      "    deprioritized_servers=[],\n",
      "    operation_id=None\n",
      ")\n",
      "   1789             raise AutoReconnect(\"server %s:%s no longer available\" % address)  # noqa: UP031\n",
      "   1790     else:\n",
      "-> 1791         server = topology.select_server(\n",
      "        topology = <Topology <TopologyDescription id: 6800bf4bb7cba4151fdd482b, topology_type: Unknown, servers: [<ServerDescription ('localhost', 54319) server_type: Unknown, rtt: None, error=AutoReconnect('localhost:54319: [Errno 61] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>\n",
      "        server_selector = Primary()\n",
      "        operation = 'find'\n",
      "        deprioritized_servers = []\n",
      "        operation_id = None\n",
      "   1792             server_selector,\n",
      "   1793             operation,\n",
      "   1794             deprioritized_servers=deprioritized_servers,\n",
      "   1795             operation_id=operation_id,\n",
      "   1796         )\n",
      "   1797     return server\n",
      "   1798 except PyMongoError as exc:\n",
      "   1799     # Server selection errors in a transaction are transient.\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/topology.py:409, in Topology.select_server(\n",
      "    self=<Topology <TopologyDescription id: 6800bf4bb7cba...MS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>,\n",
      "    selector=Primary(),\n",
      "    operation='find',\n",
      "    server_selection_timeout=None,\n",
      "    address=None,\n",
      "    deprioritized_servers=[],\n",
      "    operation_id=None\n",
      ")\n",
      "    399 def select_server(\n",
      "    400     self,\n",
      "    401     selector: Callable[[Selection], Selection],\n",
      "   (...)\n",
      "    406     operation_id: Optional[int] = None,\n",
      "    407 ) -> Server:\n",
      "    408     \"\"\"Like select_servers, but choose a random server if several match.\"\"\"\n",
      "--> 409     server = self._select_server(\n",
      "        self = <Topology <TopologyDescription id: 6800bf4bb7cba4151fdd482b, topology_type: Unknown, servers: [<ServerDescription ('localhost', 54319) server_type: Unknown, rtt: None, error=AutoReconnect('localhost:54319: [Errno 61] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>\n",
      "        selector = Primary()\n",
      "        operation = 'find'\n",
      "        server_selection_timeout = None\n",
      "        address = None\n",
      "        deprioritized_servers = []\n",
      "        operation_id = None\n",
      "    410         selector,\n",
      "    411         operation,\n",
      "    412         server_selection_timeout,\n",
      "    413         address,\n",
      "    414         deprioritized_servers,\n",
      "    415         operation_id=operation_id,\n",
      "    416     )\n",
      "    417     if _csot.get_timeout():\n",
      "    418         _csot.set_rtt(server.description.min_round_trip_time)\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/topology.py:387, in Topology._select_server(\n",
      "    self=<Topology <TopologyDescription id: 6800bf4bb7cba...MS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>,\n",
      "    selector=Primary(),\n",
      "    operation='find',\n",
      "    server_selection_timeout=None,\n",
      "    address=None,\n",
      "    deprioritized_servers=[],\n",
      "    operation_id=None\n",
      ")\n",
      "    378 def _select_server(\n",
      "    379     self,\n",
      "    380     selector: Callable[[Selection], Selection],\n",
      "   (...)\n",
      "    385     operation_id: Optional[int] = None,\n",
      "    386 ) -> Server:\n",
      "--> 387     servers = self.select_servers(\n",
      "        self = <Topology <TopologyDescription id: 6800bf4bb7cba4151fdd482b, topology_type: Unknown, servers: [<ServerDescription ('localhost', 54319) server_type: Unknown, rtt: None, error=AutoReconnect('localhost:54319: [Errno 61] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>\n",
      "        selector = Primary()\n",
      "        operation = 'find'\n",
      "        server_selection_timeout = None\n",
      "        address = None\n",
      "        operation_id = None\n",
      "    388         selector, operation, server_selection_timeout, address, operation_id\n",
      "    389     )\n",
      "    390     servers = _filter_servers(servers, deprioritized_servers)\n",
      "    391     if len(servers) == 1:\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/topology.py:294, in Topology.select_servers(\n",
      "    self=<Topology <TopologyDescription id: 6800bf4bb7cba...MS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>,\n",
      "    selector=Primary(),\n",
      "    operation='find',\n",
      "    server_selection_timeout=None,\n",
      "    address=None,\n",
      "    operation_id=None\n",
      ")\n",
      "    291     self.cleanup_monitors()\n",
      "    293 with self._lock:\n",
      "--> 294     server_descriptions = self._select_servers_loop(\n",
      "        self = <Topology <TopologyDescription id: 6800bf4bb7cba4151fdd482b, topology_type: Unknown, servers: [<ServerDescription ('localhost', 54319) server_type: Unknown, rtt: None, error=AutoReconnect('localhost:54319: [Errno 61] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>\n",
      "        server_timeout = 30\n",
      "        selector = Primary()\n",
      "        operation = 'find'\n",
      "        operation_id = None\n",
      "        address = None\n",
      "    295         selector, server_timeout, operation, operation_id, address\n",
      "    296     )\n",
      "    298     return [\n",
      "    299         cast(Server, self.get_server_by_address(sd.address)) for sd in server_descriptions\n",
      "    300     ]\n",
      "\n",
      "File ~/Library/Python/3.9/lib/python/site-packages/pymongo/synchronous/topology.py:344, in Topology._select_servers_loop(\n",
      "    self=<Topology <TopologyDescription id: 6800bf4bb7cba...MS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>>,\n",
      "    selector=Primary(),\n",
      "    timeout=30,\n",
      "    operation='find',\n",
      "    operation_id=None,\n",
      "    address=None\n",
      ")\n",
      "    333     if _SERVER_SELECTION_LOGGER.isEnabledFor(logging.DEBUG):\n",
      "    334         _debug_log(\n",
      "    335             _SERVER_SELECTION_LOGGER,\n",
      "    336             message=_ServerSelectionStatusMessage.FAILED,\n",
      "   (...)\n",
      "    342             failure=self._error_message(selector),\n",
      "    343         )\n",
      "--> 344     raise ServerSelectionTimeoutError(\n",
      "    345         f\"{self._error_message(selector)}, Timeout: {timeout}s, Topology Description: {self.description!r}\"\n",
      "    346     )\n",
      "    348 if not logged_waiting:\n",
      "    349     _debug_log(\n",
      "    350         _SERVER_SELECTION_LOGGER,\n",
      "    351         message=_ServerSelectionStatusMessage.WAITING,\n",
      "   (...)\n",
      "    357         remainingTimeMS=int(1000 * (end_time - time.monotonic())),\n",
      "    358     )\n",
      "\n",
      "ServerSelectionTimeoutError: localhost:54319: [Errno 61] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms), Timeout: 30s, Topology Description: <TopologyDescription id: 6800bf4bb7cba4151fdd482b, topology_type: Unknown, servers: [<ServerDescription ('localhost', 54319) server_type: Unknown, rtt: None, error=AutoReconnect('localhost:54319: [Errno 61] Connection refused (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>\n",
      "\n"
     ]
    }
   ],
   "execution_count": 7
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-04-17T08:43:55.410631Z",
     "start_time": "2025-04-17T08:43:55.409177Z"
    }
   },
   "source": [],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
